File: src/interpreter/lexer.ts
import { SymbolError } from "./errors";
import { Token, TokenType, SymbolToken, newToken } from "./token";
import { isspace, isdigit, isalnum } from "./util";

interface Keywords {
  [key: string]: Token;
}

/**
 * Lexer tokenizes raw input into @class{Token} objects
 */
export class Lexer {
  public static readonly reserved_keywords: Keywords = {
    true: newToken(TokenType.primitive, "true"),
    false: newToken(TokenType.primitive, "false"),
  };
  /**
   * raw input text
   */
  private text: string;
  /**
   * current tokenizing position
   */
  private position: number;
  /**
   * current character at position
   */
  private current_char: string | undefined;

  constructor(text: string) {
    this.text = text;
    this.position = -1;
  }

  private _id(): Token {
    let result = "";
    while (this.current_char && isalnum(this.current_char)) {
      result += this.current_char;
      let next = this.peek();
      if (next && isalnum(next)) {
        this.advance();
      } else {
        break;
      }
    }
    return (
      Lexer.reserved_keywords[result] || { type: TokenType.id, value: result }
    );
  }

  /**
   * increments position and updates current_char
   * sets current_char to undefined if position is out of bounds
   */
  private advance() {
    this.position++;

    if (this.position < this.text.length) {
      this.current_char = this.text.charAt(this.position);
    } else {
      this.current_char = undefined;
    }
  }

  /**
   * retrieves next character without advancing position
   * @param steps number of steps to take
   */
  public peek(steps: number = 1): string | undefined {
    if (this.position + steps >= this.text.length) {
      return undefined;
    } else {
      return this.text.charAt(this.position + steps);
    }
  }

  /**
   * retrieves the next token by ignoring whitespaces
   */
  public peekToken(): string {
    let pos = this.position + 1;
    while (pos < this.text.length && isspace(this.text.charAt(pos))) {
      pos += 1;
    }
    return this.text.charAt(pos);
  }

  /**
   * skips all whitepsace in sequence
   */
  private ignore_whitespace(advance: boolean = true) {
    if (advance) {
      // stop at next token
      while (this.current_char && isspace(this.current_char)) {
        this.advance();
      }
    } else {
      // stop before next token
      let current = this.peek();
      while (current && isspace(current)) {
        this.advance();
      }
    }
  }

  /**
   * converts numeric character sequence into a token
   */
  private tokenize_number(): number {
    let result = this.current_char || "";
    let next = this.peek();
    // capture integer
    while (next && isdigit(next)) {
      this.advance();
      result += this.current_char;
      next = this.peek();
    }

    // check if decimal point
    if (next && next == ".") {
      let tmpNext = this.peek(2);
      this.advance();
      // make sure next character is a digit, not an elipses or something like '1.'
      if (tmpNext && isdigit(tmpNext)) {
        result += this.current_char;
        next = this.peek();

        // append numbers after decimal point
        while (next && isdigit(next)) {
          this.advance();
          result += this.current_char;
          next = this.peek();
        }
      }
    }

    // convert to number and return
    return Number(result);
  }

  /**
   * retrieves next token in string
   */
  public next_token(): Token {
    this.advance();
    let token = this.tokenize();
    return token;
  }

  /**
   * compares characters and matches with associated token
   */
  private tokenize(): Token {
    // ignore all spaces
    if (this.current_char && isspace(this.current_char)) {
      this.ignore_whitespace();
    }

    if (!this.current_char) {
      // check if no more characters
      return { type: TokenType.eof, value: "eof" };
    } else if (isdigit(this.current_char)) {
      // capture numeric token
      return { type: TokenType.num, value: this.tokenize_number() };
    } else if (isalnum(this.current_char)) {
      // capture alphanumeric token
      return this._id();
    } else {
      // look for each token
      for (let type in TokenType) {
        let tok: TokenType = TokenType[type as keyof typeof TokenType];
        // if this character type is more than one, peek next
        let char = this.current_char;
        let n = TokenType[type].length;
        for (let peekStep = 1; peekStep < n; peekStep++) {
          char += this.peek(peekStep);
        }
        // if type is strictly equal to char, match and return
        if (TokenType[type] == char) {
          for (let i = 1; i < n; i++) {
            this.advance();
          }
          return newToken(tok, type.toString());
        }
      }
    }

    // token wasn't recognized
    throw new SymbolError("unexpected token: `" + this.current_char + "`");
  }
}

File: interpreter/lexer.cpp
//
// Created by Pranav Putta on 9/19/20.
//
#include "lexer.h"
#include "errors.h"

#include <utility>

using namespace std;

/**
 * constructs a new lexer from text
 * @param t
 */
Lexer::Lexer(std::string t) : text(std::move(t)), position(0) {
    if (!text.empty()) {
        current_char = text.at(0);
    } else {
        current_char = -1;
    }
}

/**
 * increment position value and set current_char to next element in text
 * if position reaches the end of text length, sets @param current_char to -1
 */
void Lexer::Advance() {
    position++;
    if (position < text.length()) {
        current_char = text.at(position);
    } else {
        current_char = -1;
    }
}

/**
 * checks for the next character in the string without advancing the current position
 * @param step size, defaults to 1
 * @return character specified step size away
 */
char Lexer::Peek(int step) {
    if (position + step < 0 || position + step >= text.length()) {
        return -1;
    } else {
        return text.at(position + step);
    }
}


/**
 * advances current character sequence position until character is not
 * a space
 */
void Lexer::IgnoreWhitespace() {
    while (isspace(current_char) && current_char != '\n') {
        Advance();
    }
}

/**
 * converts a number sequence into an integer or double token
 * @return token
 */
Token Lexer::TokenizeNumber() {
    string num;
    // aggregate consecutive numbers as integer
    while (isdigit(current_char)) {
        num += current_char;
        Advance();
    }

    // check for decimal point
    if (current_char == '.') {
        num += current_char;
        Advance();
        // make sure next character is a digit
        while (isdigit(current_char)) {
            num += current_char;
            Advance();
        }
    }
    return Token{TokenType::kNum, std::stod(num)};
}

/**
 * matches character with associated token
 * @return Token, value string if symbol, double/int if number
 */
Token Lexer::TokenizeIdentifier() {
    string str;
    while (isalnum(current_char)) {
        str += current_char;
        Advance();
    }
    int ct = Global::ReservedWords().count(str);
    return ct != 0 ? Token{Global::ReservedWords()[str], str} : Token{TokenType::kId, str};
}

/**
 * tokenizes the current character
 * @return token
 */
Token Lexer::Tokenize() {
    // ignore all whitespace
    IgnoreWhitespace();

    if (current_char == -1) {
        // if reached end of file, return eof
        return Token{TokenType::kEof, "eof"};
    } else if (isdigit(current_char)) {
        // check for digits
        return TokenizeNumber();
    } else if (isalnum(current_char)) {
        // check for alphanumeric sequence
        return TokenizeIdentifier();
    } else {
        // check for each character
        for (auto const &tok : Global::TokenMap()) {
            string tok_str = tok.second, txt(1, current_char);
            int n = tok_str.length();

            // if token to check is multiple characters, peek forward
            for (int step = 1; step < n; step++) {
                txt += Peek(step);
            }

            // if token to check and text match
            if (tok_str == txt) {
                for (int i = 0; i < n; i++) {
                    Advance();
                }
                return Token{tok.first, tok.second};
            }
        }
    }

    throw UnexpectedSymbolError(current_char, position);
}

/**
 * tokenizes current character and advances position
 * @return tokenized character
 */
Token Lexer::NextToken() {
    Token next = Tokenize();
    return next;
}

/**
 * tokenizes current character but prevents position from advancing
 * @return
 */
Token Lexer::PeekToken() {
    unsigned int posTmp = position;
    char charTmp = current_char;
    Token tok = Tokenize();
    position = posTmp;
    current_char = charTmp;
    return tok;
}




